objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+/// asset. This guarantees the properties described above (the first bit is ONE, and the 32 least
+/// significant bits of the 3rd element are all ZEROs).
+///
+/// The last element is set to the amount of the asset. This amount cannot be greater than
+/// 2^63 - 1 and thus requires 63-bits to store.
+///
+/// # Non-fungible assets
+/// The 4 elements of non-fungible assets are computed as follows:
+/// - First the asset data is hashed. This compresses an asset of an arbitrary length to 4 field
+///   elements.
+/// - The result of step 1 is hashed together with the ID of the faucet which issued the asset.
+///   This guarantees that finding two assets with the same hash is infeasible.
+/// - Lastly, the first bit of the result is set to ZERO, and the last 32 bits of the 3rd element
+///   are set to 2^31 (i.e., ONE followed by 32 ZEROs).
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Asset {
+    Fungible(FungibleAsset),
+    NonFungible(NonFungibleAsset),
+}
+
+impl From<Asset> for Word {
+    fn from(asset: Asset) -> Self {
+        use Asset::*;
+        match asset {
+            Fungible(asset) => asset.into(),
+            NonFungible(asset) => asset.into(),
+        }
+    }
+}
+
+impl From<Asset> for [u8; 32] {
+    fn from(asset: Asset) -> Self {
+        use Asset::*;
+        match asset {
+            Fungible(asset) => asset.into(),
+            NonFungible(asset) => asset.into(),
+        }
+    }
+}
+
+impl TryFrom<Word> for Asset {
+    type Error = AssetError;
+
+    fn try_from(value: Word) -> Result<Self, Self::Error> {
+        let first_bit = value[0].as_int() >> 63;
+        match first_bit {
+            0 => NonFungibleAsset::try_from(value).map(Asset::from),
+            1 => FungibleAsset::try_from(value).map(Asset::from),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl TryFrom<[u8; 32]> for Asset {
+    type Error = AssetError;
+
+    fn try_from(value: [u8; 32]) -> Result<Self, Self::Error> {
+        let first_bit = value[0] >> 7;
+        match first_bit {
+            0 => NonFungibleAsset::try_from(value).map(Asset::from),
+            1 => FungibleAsset::try_from(value).map(Asset::from),
+            _ => unreachable!(),
+        }
+    }
+}
+
+// FUNGIBLE ASSET
+// ================================================================================================
+/// A fungible asset.
+///
+/// A fungible asset consists of a faucet ID of the faucet which issued the asset as well as the
+/// asset amount. Asset amount is guaranteed to be 2^63 - 1 or smaller.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct FungibleAsset {
+    faucet_id: AccountId,
+    amount: u64,
+}
+
+impl FungibleAsset {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    /// Specifies a maximum amount value for fungible assets which can be at most a 63-bit value.
+    pub const MAX_AMOUNT: u64 = (1u64 << 63) - 1;
```

Comment:
nit/question: I thought you had mentioned you preferred using `1_u64`over `1u64`?

Reply:
Yep - my mistake! fixed!